src/ucp/proto/proto_am.inl

```diff
@@ -175,81 +201,122 @@ ucs_status_t ucp_do_am_zcopy_multi(uct_pending_req_t *self, uint8_t am_id_first,
+            /* This flag should guarantee middle stage usage if iovcnt exceeded */
+            flag_iov_mid = ((state.dt.iov.iovcnt_offset + max_iov) <
+                            state.dt.iov.iovcnt);
+        } else {
+            ucs_assert(UCP_DT_IS_CONTIG(req->send.datatype));
+        }
+
+        if (offset == 0) {
+            /* First stage */
+            ucp_dt_iov_copy_uct(ep->worker->context, iov, &iovcnt, max_iov, &state,
+                                req->send.buffer,  req->send.datatype,
+                                max_middle - hdr_size_first + hdr_size_middle,
+                                ucp_ep_md_index(ep, req->send.lane));
+
+            status = uct_ep_am_zcopy(uct_ep, am_id_first, (void*)hdr_first,
+                                     hdr_size_first, iov, iovcnt, 0,
+                                     &req->send.state.uct_comp);
+
+            UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_zcopy_first",
+                                                   iov[0].length, status);
+        } else if ((offset + max_middle < req->send.length) || flag_iov_mid) {
+            /* Middle stage */
+            ucp_dt_iov_copy_uct(ep->worker->context, iov, &iovcnt, max_iov,
+                                &state, req->send.buffer, req->send.datatype,
+                                max_middle, ucp_ep_md_index(ep, req->send.lane));
+
+            status = uct_ep_am_zcopy(uct_ep, am_id_middle, (void*)hdr_middle,
+                                     hdr_size_middle, iov, iovcnt, 0,
+                                     &req->send.state.uct_comp);
+
+            UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_zcopy_middle",
+                                                   iov[0].length, status);
+        } else {
+            /* Last stage */
+            ucp_dt_iov_copy_uct(ep->worker->context, iov, &iovcnt, max_iov, &state,
+                                req->send.buffer, req->send.datatype,
+                                req->send.length - offset,
+                                ucp_ep_md_index(ep, req->send.lane));
+
+            status = uct_ep_am_zcopy(uct_ep, am_id_last, (void*)hdr_middle,
+                                     hdr_size_middle, iov, iovcnt, 0,
+                                     &req->send.state.uct_comp);
+            UCS_PROFILE_REQUEST_EVENT_CHECK_STATUS(req, "am_zcopy_last",
+                                                   iov[0].length, status);
+            if (status == UCS_OK) {
+                complete(req, UCS_OK);
+                return UCS_OK;
+            }
+            ucp_request_send_state_advance(req, &state,
+                                           UCP_REQUEST_SEND_PROTO_ZCOPY_AM,
+                                           status);
+            if (!UCS_STATUS_IS_ERR(status)) {
+                return UCS_OK;
+            }
+        }
+
+        if (status == UCS_ERR_NO_RESOURCE) {
+            if (req->send.lane != req->send.pend_add_lane) {
```

Comment:
can we make common function for am_bcopy and am_zcopy which does these checks?

Reply:
shifting it to separate function will make it complicated: here is used ```continue``` to jump to loop iteration which is impossible from function, additionally there are different reaction on NO_RES error (return status or packed bytes)